PR: https://github.com/odoo/odoo/pull/257098

From: 9937f092b5ea73dfa3d56e7dc034ab83a199397f
From: Haroune Hassine (Hahas)
Date: 2026-06-19 16:06:17

Breaking data model changes scores: del:3 + add:1, change matches:
-    mailing_filter_id = fields.Many2one(
-    mailing_filter_domain = fields.Char('Favorite filter domain', related='mailing_filter_id.mailing_domain')
-    mailing_filter_count = fields.Integer('# Favorite Filters', compute='_compute_mailing_filter_count')
+    mailing_filter_ids = fields.Many2many(

Total Changes: 248

[IMP] mass_mailing: add dynamic mailing lists

PURPOSE

The purpose of this commit is to add dynamic mailing list
alongside the exsiting manual/static lists. Dynamic lists are list
defined by a target mailing model and a domain.

This will make it easier for users to re-use previously created
domains, instead of having to set the domain for the selected model
each time they create a new mailing.

This feature comes as an extension of the `mailing.filter` (Favorite
Filters), with better UX/UI.

We also added:
* The possibility for the user to save a crafted domain in the
domain widget as a dynamic list.
* A label that shows the total recipients count of the selected
domains, for mailing models other that `mailing.list`.

The `mailing_filter` widget, which handles the favorite filters of
the mailings, has been replaced by the m2m tag using the dynamic
lists (`mailing.filter` model), added in a previous commit.
Therefore, the `mailing_filter` widget is no longer used in the
code and should be removed.

SPECIFICATIONS

Dynamic Lists
*-----

The user can create a dynamic list `mailing.filter` to save a
specific domain to be re-used later.

Some ready dynamic list templates are provided for the user when
creating a new dynamic list.

When creating a new mailing with a target model other than mailing
list, the user can choose to set the domain manually or choose
from the pre-fabricated dynamic lists (filters).

When choosing dynamic lists, the final mailing domain will be the
union of all the domains of the provided lists.

Domain Saving
*-----

* When the user creates a custom domain using the domain widget,
the `save as..` button will appear.
* When clicking on the save button, a mailing.filter form dialog
will pop up with the created domain set as default domain.
* When saving the domain as a new dynamic list, that list will be
set as the dynamic list of the current mailing.
(Note: the working mailing should have been saved before saving a
domain)

Recipients Count
*-----

Nothing too special apart from adding the total recipients count
when selecting dynamic lists as recipients, or when setting a domain
for models other than `mailing.list`.

task-5358279

Part-of: odoo/odoo#257098
Related: odoo/enterprise#114690
Related: odoo/upgrade#10095
Signed-off-by: Warnon Aurélien (awa) <awa@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mass_mailing/models/mailing.py
+++ b/addons/mass_mailing/models/mailing.py
@@ -178,13 +178,11 @@ class MailingMailing(models.Model):
         'Use Exclusion List', default=True, copy=False, store=True,
         readonly=False, compute='_compute_use_exclusion_list',
         help='Prevent sending messages to blacklisted contacts. Disable only when absolutely necessary.')
+    recipients_count = fields.Integer("# Of Recipients", compute='_compute_recipients_count')
     # Mailing Filter
-    mailing_filter_id = fields.Many2one(
-        'mailing.filter', string='Favorite Filter',
-        compute='_compute_mailing_filter_id', readonly=False, store=True,
-        domain="[('mailing_model_name', '=', mailing_model_name)]")
-    mailing_filter_domain = fields.Char('Favorite filter domain', related='mailing_filter_id.mailing_domain')
-    mailing_filter_count = fields.Integer('# Favorite Filters', compute='_compute_mailing_filter_count')
+    mailing_filter_ids = fields.Many2many(
+        'mailing.filter', 'mail_mass_mailing_filter_rel', 'mailing_mailing_id', 'mailing_filter_id', compute='_compute_mailing_filter_ids',
+        readonly=False, store=True, string="Dynamic Lists", domain="[('mailing_model_name', '=', mailing_model_name)]")
     # A/B Testing
     ab_testing_completed = fields.Boolean(related='campaign_id.ab_testing_completed')
     ab_testing_description = fields.Html('A/B Testing Description', compute="_compute_ab_testing_description")
@@ -264,13 +262,16 @@ class MailingMailing(models.Model):
 
         return super()._order_field_to_sql(table, field_expr, direction, nulls)
 
-    @api.constrains('mailing_model_id', 'mailing_filter_id')
+    @api.constrains('mailing_model_id', 'mailing_filter_ids')
     def _check_mailing_filter_model(self):
-        """Check that if the favorite filter is set, it must contain the same recipient model as mailing"""
+        """Check that if the dynamic lists are set, they must contain the same recipient model as the mailing"""
         for mailing in self:
-            if mailing.mailing_filter_id and mailing.mailing_model_id != mailing.mailing_filter_id.mailing_model_id:
+            if any(
+                mailing.mailing_model_id != mailing_filter_id.mailing_model_id
+                for mailing_filter_id in mailing.mailing_filter_ids
+            ):
                 raise ValidationError(
-                    _("The saved filter targets different recipients and is incompatible with this mailing.")
+                    _("The saved filters target different recipients and are incompatible with this mailing.")
                 )
 
     @api.depends('campaign_id.ab_testing_winner_mailing_id')
@@ -446,15 +447,6 @@ class MailingMailing(models.Model):
             elif mailing.reply_to_mode == 'update':
                 mailing.reply_to = False
 
-    @api.depends('mailing_model_id', 'mailing_domain')
-    def _compute_mailing_filter_count(self):
-        filter_data = self.env['mailing.filter']._read_group([
-            ('mailing_model_id', 'in', self.mailing_model_id.ids)
-        ], ['mailing_model_id'], ['__count'])
-        mapped_data = {mailing_model.id: count for mailing_model, count in filter_data}
-        for mailing in self:
-            mailing.mailing_filter_count = mapped_data.get(mailing.mailing_model_id.id, 0)
-
     @api.depends('mailing_model_id')
     def _compute_mailing_model_real(self):
         for mailing in self:
@@ -470,13 +462,13 @@ class MailingMailing(models.Model):
         self.mailing_on_mailing_list = False
         self.filtered(lambda m: m.mailing_model_id == mailing_list_model_id).mailing_on_mailing_list = True
 
-    @api.depends('mailing_model_id', 'contact_list_ids', 'mailing_type', 'mailing_filter_id')
+    @api.depends('mailing_model_id', 'contact_list_ids', 'mailing_type', 'mailing_filter_ids', 'mailing_filter_ids.mailing_domain')
     def _compute_mailing_domain(self):
         for mailing in self:
             if not mailing.mailing_model_id:
                 mailing.mailing_domain = ''
-            elif mailing.mailing_filter_id:
-                mailing.mailing_domain = mailing.mailing_filter_id.mailing_domain
+            elif mailing.mailing_filter_ids:
+                mailing.mailing_domain = repr(Domain.OR(literal_eval(mailing_filter.mailing_domain) for mailing_filter in mailing.mailing_filter_ids))
             else:
                 mailing.mailing_domain = repr(mailing._get_default_mailing_domain() or [])
 
@@ -490,9 +482,9 @@ class MailingMailing(models.Model):
         ).use_exclusion_list = True
 
     @api.depends('mailing_model_name')
-    def _compute_mailing_filter_id(self):
+    def _compute_mailing_filter_ids(self):
         for mailing in self:
-            mailing.mailing_filter_id = False
+            mailing.mailing_filter_ids = [fields.Command.clear()]
 
     @api.depends('schedule_type')
     def _compute_schedule_date(self):
@@ -551,6 +543,16 @@ class MailingMailing(models.Model):
     def _get_ab_testing_description_modifying_fields(self):
         return ['ab_testing_enabled', 'ab_testing_pc', 'ab_testing_schedule_datetime', 'ab_testing_winner_selection', 'campaign_id']
 
+    @api.depends('mailing_domain', 'contact_list_ids')
+    def _compute_recipients_count(self):
+        """Calculate the total number of recipients in the mailing.
+        (Opted out contacts are not excluded from the count)"""
+        for mailing in self:
+            if mailing.mailing_domain:
+                mailing.recipients_count = self.env[mailing.mailing_model_real].search_count(mailing._get_recipients_domain())
+            elif mailing.contact_list_ids:
+                mailing.recipients_count = sum(contact_list.contact_count for contact_list in mailing.contact_list_ids)
+
     # ------------------------------------------------------
     # ORM
     # ------------------------------------------------------
@@ -737,19 +739,19 @@ class MailingMailing(models.Model):
         return action
 
     def action_view_clicked(self):
-        return self.env['link.tracker.click']._action_view_mailing_statistics(Domain('mass_mailing_id', '=', self.id))
+        return self.env['link.tracker.click']._action_view_mailing_statistics(Domain('mass_mailing_id', 'in', self.ids), group_by_field='email', graph_mode=True)
 
     def action_view_opened(self):
-        return self.env['mailing.trace']._action_view_mailing_statistics_filtered(Domain('mass_mailing_id', '=', self.id), view_filter='open')
+        return self.env['mailing.trace']._action_view_mailing_statistics_filtered(Domain('mass_mailing_id', 'in', self.ids), view_filter='open')
 
     def action_view_replied(self):
-        return self.env['mailing.trace']._action_view_mailing_statistics_filtered(Domain('mass_mailing_id', '=', self.id), view_filter='reply')
+        return self.env['mailing.trace']._action_view_mailing_statistics_filtered(Domain('mass_mailing_id', 'in', self.ids), view_filter='reply')
 
     def action_view_bounced(self):
-        return self.env['mailing.trace']._action_view_mailing_statistics_filtered(Domain('mass_mailing_id', '=', self.id), view_filter='bounce')
+        return self.env['mailing.trace']._action_view_mailing_statistics_filtered(Domain('mass_mailing_id', 'in', self.ids), view_filter='bounce')
 
     def action_view_delivered(self):
-        return self.env['mailing.trace']._action_view_mailing_statistics_filtered(Domain('mass_mailing_id', '=', self.id), view_filter='delivered')
+        return self.env['mailing.trace']._action_view_mailing_statistics_filtered(Domain('mass_mailing_id', 'in', self.ids), view_filter='delivered')
 
     def action_import_mailing_contacts(self):
         """Display the mailing contact import-by-paste wizard for this mailing's contact lists"""

--- a/addons/mass_mailing/models/mailing_contact.py
+++ b/addons/mass_mailing/models/mailing_contact.py
@@ -9,6 +9,7 @@ from odoo import _, api, fields, models, tools
 from odoo.exceptions import UserError
 from odoo.fields import Domain
 from odoo.tools import clean_context
+from odoo.tools.float_utils import float_round
 
 MAX_FROM_PARTNERS = 500
 
@@ -54,6 +55,16 @@ class MailingContact(models.Model):
         compute='_compute_opt_out', search='_search_opt_out',
         help='Opt out flag for a specific mailing list. '
              'This field should not be used in a view without a unique and active mailing list context.')
+    mailing_count = fields.Integer('Number of Mailing', compute='_compute_statistics')
+    received_ratio = fields.Float('Received Ratio', compute='_compute_statistics')
+    opened_ratio = fields.Float('Opened Ratio', compute='_compute_statistics')
+    replied_ratio = fields.Float('Replied Ratio', compute='_compute_statistics')
+    clicks_ratio = fields.Float('Clicks Ratio', compute="_compute_clicks_ratio")
+    # Datetimes that mirror the onces in mailing_trace for the corresponding contact.
+    # They get updated from the mailing.trace model.
+    last_opened_datetime = fields.Datetime('Last Opened On')
+    last_clicked_datetime = fields.Datetime('Last Clicked On')
+    last_replied_datetime = fields.Datetime('Last Replied On')
 
     partner_id = fields.Many2one('res.partner', string='Contact', index='btree_not_null')
 
@@ -109,6 +120,44 @@ class MailingContact(models.Model):
             return self.env.context['default_list_ids'][0]
         return None
 
+    def _compute_clicks_ratio(self):
+        grouped_traces = dict(self.env['mailing.trace'].sudo()._read_group(
+            domain=[('model', '=', 'mailing.contact'), ('res_id', 'in', self.ids)],
+            groupby=[('res_id')],
+            aggregates=[('id:count_distinct')]
+        ))
+        grouped_clicks = dict(self.env['link.tracker.click'].sudo()._read_group(
+            domain=[('mailing_trace_id.model', '=', 'mailing.contact'), ('mailing_trace_id.res_id', 'in', self.ids)],
+            groupby=[('mailing_trace_id.res_id')],
+            aggregates=[('mailing_trace_id:count_distinct')]
+        ))
+        for contact in self:
+            contact.clicks_ratio = float_round(100 * grouped_clicks.get(contact.id, 0) / grouped_traces.get(contact.id, 1), precision_digits=2)
+
+    def _compute_statistics(self):
+        """ Compute the mailing statistics for the mailing contact """
+        result = self.env["mailing.trace"].sudo()._read_group(
+            [('model', '=', 'mailing.contact'), ('res_id', 'in', self.ids)],
+            ['res_id', 'trace_status'],
+            ['__count'])
+
+        result_per_contact = defaultdict(lambda: defaultdict(int))
+        for res_id, trace_status, count in result:
+            result_per_contact[res_id][trace_status] = count
+
+        for contact in self:
+            line = result_per_contact[contact.id]
+            expected = sum(line.values())
+            delivered = line['sent'] + line['open'] + line['reply']
+            opened = line['open'] + line['reply']
+            failed = line['error'] + line['bounce']
+            total = (expected - line['cancel']) or 1
+            total_no_error = (expected - line['cancel'] - failed) or 1
+            contact.mailing_count = delivered
+            contact.received_ratio = float_round(100.0 * delivered / total, precision_digits=2)
+            contact.opened_ratio = float_round(100.0 * opened / total_no_error, precision_digits=2)
+            contact.replied_ratio = float_round(100.0 * line['reply'] / total_no_error, precision_digits=2)
+
     @api.model_create_multi
     def create(self, vals_list):
         """ Synchronize default_list_ids (currently used notably for computed
@@ -195,7 +244,7 @@ class MailingContact(models.Model):
         return action
 
     def action_open_base_import(self):
-        """Open the base import wizard to import mailing list contacts with a xlsx file."""
+        """Open the base import wizard to import mailing contacts with a xlsx file."""
 
         return {
             'type': 'ir.actions.client',
@@ -207,10 +256,37 @@ class MailingContact(models.Model):
             },
         }
 
+    def action_view_mailings(self):
+        action = self.env["ir.actions.actions"]._for_xml_id('mass_mailing.mailing_mailing_action_mail')
+        traces_grouped = dict(self.env['mailing.trace']._read_group(
+            [('model', '=', 'mailing.contact'), ('res_id', 'in', self.ids), ('sent_datetime', '!=', False)],
+            ['mass_mailing_id'],
+            ['__count']
+        ))
+        mailing_ids = [m.id for m in traces_grouped]
+        action['domain'] = [('id', 'in', mailing_ids)]
+        return action
+
+    def action_view_received(self):
+        domain = Domain.AND([Domain('model', '=', 'mailing.contact'), Domain('res_id', 'in', self.ids)])
+        return self.env['mailing.trace'].with_context({'search_default_filter_delivered': 1, 'search_default_group_sent_date': 1})._action_view_mailing_statistics_filtered(domain, 'delivered')
+
+    def action_view_opened(self):
+        domain = Domain.AND([Domain('model', '=', 'mailing.contact'), Domain('res_id', 'in', self.ids)])
+        return self.env['mailing.trace'].with_context({'search_default_filter_opened': 1})._action_view_mailing_statistics_filtered(domain, 'open')
+
+    def action_view_replied(self):
+        domain = Domain.AND([Domain('model', '=', 'mailing.contact'), Domain('res_id', 'in', self.ids)])
+        return self.env['mailing.trace'].with_context({'search_default_filter_replied': 1})._action_view_mailing_statistics_filtered(domain, 'reply')
+
+    def action_view_clicked(self):
+        domain = [('mailing_trace_id.model', '=', 'mailing.contact'), ('mailing_trace_id.res_id', 'in', self.ids)]
+        return self.env['link.tracker.click']._action_view_mailing_statistics(domain)
+
     @api.model
     def get_import_templates(self):
         return [{
-            'label': _('Template for Mailing List Contacts'),
+            'label': _('Template for Mailing Contacts'),
             'template': '/mass_mailing/static/xls/mailing_contact.xls'
         }]
 

--- a/addons/mass_mailing/models/mailing_filter.py
+++ b/addons/mass_mailing/models/mailing_filter.py
@@ -4,23 +4,26 @@
 from ast import literal_eval
 
 from odoo import api, fields, models, _
-from odoo.exceptions import ValidationError
+from odoo.exceptions import ValidationError, UserError
 
 
 class MailingFilter(models.Model):
-    """ This model stores mass mailing or marketing campaign domain as filters
-    (quite similar to 'ir.filters' but dedicated to mailing apps). Frequently
-    used domains can be reused easily. """
+    """ This model stores mass mailing or marketing campaign domain as dynamic
+    lists (quite similar to 'ir.filters' but dedicated to mailing apps).
+    Frequently used domains can be reused easily."""
     _name = 'mailing.filter'
-    _description = 'Mailing Favorite Filter'
+    _description = 'Mailing Dynamic List'
     _order = 'create_date DESC'
 
     # override create_uid field to display default value while creating filter from 'Configuration' menus
     create_uid = fields.Many2one('res.users', 'Saved by', index=True, readonly=True, default=lambda self: self.env.user)
-    name = fields.Char(string='Filter Name', required=True)
-    mailing_domain = fields.Char(string='Filter Domain', required=True)
-    mailing_model_id = fields.Many2one('ir.model', string='Recipients Model', required=True, ondelete='cascade')
+    name = fields.Char(string='List Name', required=True)
+    active = fields.Boolean(default=True)
+    color = fields.Integer(string='Color', default=0)
+    mailing_domain = fields.Char(string='Filter Domain', required=True, default="[]")
+    mailing_model_id = fields.Many2one('ir.model', string='Recipients Model', required=True, ondelete='cascade', domain=[('is_mailing_enabled', '=', True)])
     mailing_model_name = fields.Char(string='Recipients Model Name', related='mailing_model_id.model')
+    mailing_count = fields.Integer('Number of Mailing', compute='_compute_mailing_count')
 
     @api.constrains('mailing_domain', 'mailing_model_id')
     def _check_mailing_domain(self):
@@ -33,3 +36,69 @@ class MailingFilter(models.Model):
                     raise ValidationError(
                         _("The filter domain is not valid for this recipients.")
                     )
+
+    # ------------------------------------------------------
+    # COMPUTE
+    # ------------------------------------------------------
+
+    def _compute_mailing_count(self):
+        groups = dict(
+            self.env['mailing.mailing']._read_group(
+            domain=[('mailing_filter_ids', 'in', self.ids)],
+            groupby=['mailing_filter_ids'],
+            aggregates=['__count']))
+        for mailing_filter in self:
+            mailing_filter.mailing_count = groups.get(mailing_filter, 0)
+
+    # ------------------------------------------------------
+    # ACTIONS
+    # ------------------------------------------------------
+
+    def action_view_recipients(self):
+        """Redirect to the list view of the model of the selected dynamic
+        list, with the domain of the latter set as the domain of that view.
+        If the model is `mailing.contact`, the `New`, `Paste` and `Upload`
+        buttons will be hidden."""
+        self.ensure_one()
+        if self.mailing_model_name == 'mailing.contact':
+            views = [(self.env.ref('mass_mailing.mailing_contact_view_list_from_filter').id, 'list'), (False, 'form')]
+        else:
+            views = [(False, 'list'), (False, 'form')]
+
+        return {
+            'type': 'ir.actions.act_window',
+            'name': self.mailing_model_id.name,
+            'res_model': self.mailing_model_name,
+            'views': views,
+            'target': 'current',
+            'domain': literal_eval(self.mailing_domain),
+            'context': {'create': False},
+            'help': self.env._("""
+                <p class="o_view_nocontent_smiling_face">No record currently matches your rules!</p>
+            """),
+        }
+
+    def action_view_mailings(self):
+        action = self.env["ir.actions.actions"]._for_xml_id('mass_mailing.mailing_mailing_action_mail')
+        action['domain'] = [('mailing_filter_ids', 'in', self.ids)]
+        return action
+
+    def action_send_mailing(self):
+        """Open the mailing form view, with the current model & domain set as the mailing model & mailing domain.
+        respectively"""
+        if not self.mailing_model_id.is_mailing_enabled:
+            raise UserError(_("Cannot use the model %s to send a mailing. Only mailing models can be used.", self.mailing_model_id.name))
+        action = self.env["ir.actions.actions"]._for_xml_id('mass_mailing.mailing_mailing_action_mail')
+        action.update({
+            'context': {
+                'default_mailing_filter_ids': self.ids,
+                'default_mailing_model_id': self.mailing_model_id.id},
+            'views': [(False, 'form')]
+        })
+        return action
+
+    def action_duplicate(self):
+        self.ensure_one()
+        if self.copy():
+            return {'type': 'ir.actions.client', 'tag': 'soft_reload'}
+        return False

--- a/addons/mass_mailing/models/mailing_trace.py
+++ b/addons/mass_mailing/models/mailing_trace.py
@@ -6,6 +6,7 @@ from markupsafe import Markup
 from odoo import api, fields, models, _
 from odoo.fields import Domain
 
+
 class MailingTrace(models.Model):
     """ MailingTrace models the statistics collected about emails. Those statistics
     are stored in a separated model and table to avoid bloating the mail_mail table
@@ -74,8 +75,8 @@ class MailingTrace(models.Model):
     medium_id = fields.Many2one(related='mass_mailing_id.medium_id')
     source_id = fields.Many2one(related='mass_mailing_id.source_id')
     # document
-    model = fields.Char(string='Document model', required=True)
-    res_id = fields.Many2oneReference(string='Document ID', model_field='model')
+    model = fields.Char(string='Document model', required=True, index=True)
+    res_id = fields.Many2oneReference(string='Document ID', model_field='model', index=True)
     # campaign data
     mass_mailing_id = fields.Many2one('mailing.mailing', string='Mailing', index=True, ondelete='cascade')
     campaign_id = fields.Many2one(
@@ -161,16 +162,34 @@ class MailingTrace(models.Model):
         that are not already opened or replied. """
         traces = self + (self.search(domain) if domain else self.env['mailing.trace'])
         traces.filtered(lambda t: t.trace_status not in ('open', 'reply')).write({'trace_status': 'open', 'open_datetime': fields.Datetime.now()})
+        if contact_traces := traces.filtered(lambda trace: trace.model == 'mailing.contact'):
+            # Apply side effects on `mailing.contact`
+            contacts = self.env['mailing.contact'].search([('id', 'in', contact_traces.mapped('res_id'))])
+            contacts.write({'last_opened_datetime': fields.Datetime.now()})
         return traces
 
     def set_clicked(self, domain=None):
         traces = self + (self.search(domain) if domain else self.env['mailing.trace'])
         traces.write({'links_click_datetime': fields.Datetime.now()})
+        if contact_traces := traces.filtered(lambda trace: trace.model == 'mailing.contact'):
+            # Apply side effects on `mailing.contact`
+            contacts = self.env['mailing.contact'].search([('id', 'in', contact_traces.mapped('res_id'))])
+            contacts.write({
+                'last_clicked_datetime': fields.Datetime.now(),
+                'last_opened_datetime': fields.Datetime.now()}
+            )
         return traces
 
     def set_replied(self, domain=None):
         traces = self + (self.search(domain) if domain else self.env['mailing.trace'])
         traces.write({'trace_status': 'reply', 'reply_datetime': fields.Datetime.now()})
+        if contact_traces := traces.filtered(lambda trace: trace.model == 'mailing.contact'):
+            # Apply side effects on `mailing.contact`
+            contacts = self.env['mailing.contact'].search([('id', 'in', contact_traces.mapped('res_id'))])
+            contacts.write({
+                'last_replied_datetime': fields.Datetime.now(),
+                'last_opened_datetime': fields.Datetime.now()
+            })
         return traces
 
     def set_bounced(self, domain=None, bounce_message=False):
